### PR TITLE
UNR-4098 Cross-server event tracing 

### PR DESCRIPTION
**Pull request by improbable-danny**
#### Description
Implementation for cross-server RPC and property tracing. 
Also includes a helper blueprint function to reduce blueprint code for simple events

#### Tests
Worker tooling and unit tests passing. Follow-up unit tests in UNR-4465

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracerUserInterface.cpp [hunk 1]
```diff
@@ -11,6 +11,27 @@
 
 DEFINE_LOG_CATEGORY(LogSpatialEventTracerUserInterface);
 
+namespace
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracerUserInterface.cpp:14]: Added some convenience functions/blueprints here as a driveby

File: SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/SpatialTraceUniqueId.h [hunk 2]
```diff
@@ -12,19 +12,21 @@ class UFunction;
 
 namespace SpatialGDK
 {
-struct EventTraceUniqueId
+class EventTraceUniqueId
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/SpatialTraceUniqueId.h:15]: Had a bug where I passed an integer into EventTraceUniqueId directly which was not intentional behavior

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp [hunk 3]
```diff
@@ -218,8 +218,9 @@ void SpatialEventTracer::StreamDeleter::operator()(Io_Stream* StreamToDestroy) c
 
 void SpatialEventTracer::BeginOpsForFrame()
 {
-	EntityComponentSpanIds.Empty(
-		EntityComponentSpanIds.Num()); // Reset all entries. It is assumed all entries are consumed during processing.
+	// Reset all entries. It is assumed all entries are consumed during processing.
+	EntityComponentSpanIds.Empty(EntityComponentSpanIds.Num());
+	RequestSpanIds.Empty(RequestSpanIds.Num());
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp:223]: RequestSpanIds is introduced to track incoming command requests

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/EventTracingTest.cpp [hunk 4]
```diff
@@ -21,7 +21,7 @@ const FName AEventTracingTest::ReceivePropertyUpdateEventName = "unreal_gdk.rece
 const FName AEventTracingTest::PushRPCEventName = "unreal_gdk.push_rpc";
 const FName AEventTracingTest::ProcessRPCEventName = "unreal_gdk.process_rpc";
 const FName AEventTracingTest::ComponentUpdateEventName = "unreal_gdk.component_update";
-const FName AEventTracingTest::MergeComponentUpdateEventName = "unreal_gdk.merge_component_update";
+const FName AEventTracingTest::MergeComponentUpdateEventName = "unreal_gdk.merge_component_update"; // TODO: Fix this test!
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/EventTracingTest.cpp:24]: Fixing this as part of the ongoing unit test work

File: SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp [hunk 5]
```diff
@@ -24,8 +28,20 @@ void CrossServerRPCSender::SendCommand(const FUnrealObjectRef InTargetObjectRef,
...
+	RPCPayload::WriteToSchemaObject(CommandRequest.GetRequestObject(), InTargetObjectRef.Offset, Info.Index, UniqueRPCId,
 									InPayload.PayloadData.GetData(), InPayload.PayloadData.Num());
+
+	FSpatialGDKSpanId SpanId;
+	if (EventTracer)
+	{
+		SpanId = EventTracer->TraceEvent(
+			FSpatialTraceEventBuilder::CreateSendCrossServerRPC(
+				TargetObject, Function, EventTraceUniqueId::GenerateForCrossServerRPC(InTargetObjectRef.Entity, UniqueRPCId)),
+			EventTracer->GetFromStack().GetConstId(), 1);
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp:42]: Cool!

EventTracer->GetFromStack() allows us to latch onto user events right?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/RPCExecutor.cpp [hunk 6]
```diff
@@ -10,7 +11,7 @@
 
 namespace SpatialGDK
 {
-RPCExecutor::RPCExecutor(class USpatialNetDriver* InNetDriver)
+RPCExecutor::RPCExecutor(USpatialNetDriver* InNetDriver)
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/RPCExecutor.cpp:14]: Uber Nit: It might be preferential to pass in the event tracer? Might not be the case now but it might make testing easier in some far off universe.

Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/RPCExecutor.cpp:14]: yeah that makes sense!

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/EventTracingTest.cpp [hunk 7]
```diff
@@ -128,29 +150,31 @@ void AEventTracingTest::GatherData()
 		return A.CreationTime > B.CreationTime;
 	});
 
-	bool bFoundClient = false;
-	bool bFoundWorker = false;
+	FPlatformProcess::Sleep(1); // Worker bug means file may not be flushed by the OS (WRK-2396)
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/EventTracingTest.cpp:153]: Alternatively we could just close the stream, I think I prefer that over a sleep

Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/EventTracingTest.cpp:153]: In the future we can just use the flush API worker provide

File: SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp (hunk 5 above)
Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp:42]: Yep! They will be pushed here around the delegate call (it uses the same RPC tracking code for regular RPCs)

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 8]
```diff
@@ -173,22 +173,20 @@ void USpatialReceiver::OnCommandResponse(const Worker_Op& Op)
...
-
 		if (EventTracer != nullptr)
 		{
 			EventTracer->TraceEvent(
 				FSpatialTraceEventBuilder::CreateReceiveCommandResponse(TEXT("SERVER_WORKER_FORWARD_SPAWN_REQUEST_COMMAND"), RequestId),
-				Op.span_id, 1);
+				EventTracer->GetSpanForRequestId(CommandResponseOp.request_id).GetConstId(), 1);
 		}
-
+		NetDriver->PlayerSpawner->ReceiveForwardPlayerSpawnResponse(CommandResponseOp);
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:189]: Interested if the order change matters here?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/EventTracingTest.cpp [hunk 9]
```diff
@@ -128,29 +150,31 @@ void AEventTracingTest::GatherData()
...
-		if (!bFoundWorker && FileCreation.FilePath.Contains("UnrealWorker"))
+		if (FoundWorker != 2 && FileCreation.FilePath.Contains("UnrealWorker"))
 		{
 			GatherDataFromFile(FileCreation.FilePath);
-			bFoundWorker = true;
+			FoundWorker++;
 		}
 
-		if (bFoundClient && bFoundWorker)
+		if (FoundClient == 1 && FoundWorker == 2)
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/EventTracingTest.cpp:171]: Hmm maybe we make these numbers configurable to the tests? ClientsRequired = 1  WorkersRequired = 2

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/EventTracingTest.cpp [hunk 10]
```diff
@@ -128,29 +150,31 @@ void AEventTracingTest::GatherData()
...
 		{
 			break;
 		}
 	}
 
-	if (!bFoundClient || !bFoundWorker)
+	if (FoundClient != 1 || FoundWorker != 2)
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/EventTracingTest.cpp:177]: This suggests we need to run in multi worker. That should not have to be the case.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/ProcessCrossServerRPCEventTracingTest.cpp [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+		EventsTested++;
+
+		if (EventName == ApplyCrossServerRPCName)
+		{
+			if (!CheckEventTraceCause(SpanIdString, { ReceiveCrossServerRPCName }, true))
+			{
+				EventsFailed++;
+			}
+		}
+		// 		else // EventName == ReceiveRPCEventName
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/ProcessCrossServerRPCEventTracingTest.cpp:37]: Still need the commented code?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/UserProcessCrossServerPropertyEventTracingTest.cpp [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+void AUserProcessCrossServerPropertyEventTracingTest::FinishEventTraceTest()
+{
+	int EventsTested = 0;
+	int EventsFailed = 0;
+	for (const auto& Pair : TraceEvents)
+	{
+		const FString& SpanIdString = Pair.Key;
+		const FName& EventName = Pair.Value;
+
+		if (EventName != UserReceiveCrossServerPropertyEventName)
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/UserProcessCrossServerPropertyEventTracingTest.cpp:23]: Does this not want to be ReceivePropertyUpdateEventName?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/UserProcessCrossServerRPCEventTracingTest.cpp [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+void AUserProcessCrossServerRPCEventTracingTest::FinishEventTraceTest()
+{
+	int EventsTested = 0;
+	int EventsFailed = 0;
+	for (const auto& Pair : TraceEvents)
+	{
+		const FString& SpanIdString = Pair.Key;
+		const FName& EventName = Pair.Value;
+
+		if (EventName != UserReceiveCrossServerRPCEventName)
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/UserProcessCrossServerRPCEventTracingTest.cpp:23]: Again I think we want to use ApplyCrossServerRPCName here?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracerUserInterface.cpp [hunk 14]
```diff
@@ -11,6 +11,27 @@
...
+	for (const FUserSpanId& UserSpanIdCause : Causes)
+	{
+		if (!UserSpanIdCause.IsValid())
+		{
+			UE_LOG(LogSpatialEventTracerUserInterface, Warning,
+				   TEXT("USpatialEventTracerUserInterface::CreateSpanIdWithCauses - Invalid input cause"));
+			continue;
+		}
+
+		FSpatialGDKSpanId CauseSpanId = SpatialGDK::SpatialEventTracer::UserSpanIdToGDKSpanId(UserSpanIdCause);
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracerUserInterface.cpp:28]: This defo cant fail?

Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracerUserInterface.cpp:28]: Nah, only if `¬UserSpanIdCause.IsValid` but that case is already checked above ^

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 8 above)
Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:189]: Just for consistency because we usually create the event before performing the action but it doesn't matter much here

File: SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp [hunk 15]
```diff
@@ -24,8 +28,20 @@ void CrossServerRPCSender::SendCommand(const FUnrealObjectRef InTargetObjectRef,
 
 	CommandRequest CommandRequest(SpatialConstants::SERVER_TO_SERVER_COMMAND_ENDPOINT_COMPONENT_ID,
 								  SpatialConstants::UNREAL_RPC_ENDPOINT_COMMAND_ID);
-	RPCPayload::WriteToSchemaObject(CommandRequest.GetRequestObject(), InTargetObjectRef.Offset, Info.Index, FMath::RandHelper(INT_MAX),
+
+	uint64 UniqueRPCId = FMath::RandHelper(INT_MAX);
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp:32]: Curiosity, was 64bit chosen because it has enough chance of not colliding and not affecting egress, versus let's say a 128 bit GUID?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp (hunk 5 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp:42]: Usually prefer adding `/* VarName */` to an undeclared var where possible, lil easier to maintain. Like here would be `/* LinearTraceId*/`

File: SpatialGDK/Source/SpatialGDK/Private/Tests/CrossServerRPCHandlerTest.cpp [hunk 16]
```diff
@@ -103,7 +103,7 @@ CROSSSERVERRPCHANDLER_TEST(GIVEN_rpc_WHEN_resolved_and_no_queue_THEN_execute)
 
 	// No queue: Handler should be able to execute it
 	ViewCoordinator Coordinator(MoveTemp(ConnHandler), nullptr, ComponentSetData);
-	CrossServerRPCHandler Handler(Coordinator, MakeUnique<MockRPCExecutor>());
+	CrossServerRPCHandler Handler(Coordinator, MakeUnique<MockRPCExecutor>(), nullptr);
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDK/Private/Tests/CrossServerRPCHandlerTest.cpp:106]: Same here with the `/* VarName */`, your choice

File: SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCHandler.cpp [hunk 17]
```diff
@@ -2,16 +2,20 @@
...
+#include "Interop/Connection/SpatialTraceEventBuilder.h"
 #include "Interop/Connection/SpatialWorkerConnection.h"
 #include "SpatialGDKSettings.h"
 
 DEFINE_LOG_CATEGORY(LogCrossServerRPCHandler);
 
 namespace SpatialGDK
 {
-CrossServerRPCHandler::CrossServerRPCHandler(ViewCoordinator& InCoordinator, TUniquePtr<RPCExecutorInterface> InRPCExecutor)
+CrossServerRPCHandler::CrossServerRPCHandler(ViewCoordinator& InCoordinator, TUniquePtr<RPCExecutorInterface> InRPCExecutor,
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCHandler.cpp:14]: Could we default the `InEventTracer` to nullptr here?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 18]
```diff
@@ -173,22 +173,20 @@ void USpatialReceiver::OnCommandResponse(const Worker_Op& Op)
...
 	else if (ComponentId == SpatialConstants::SERVER_WORKER_COMPONENT_ID)
 	{
-		NetDriver->PlayerSpawner->ReceiveForwardPlayerSpawnResponse(CommandResponseOp);
-
 		if (EventTracer != nullptr)
 		{
 			EventTracer->TraceEvent(
 				FSpatialTraceEventBuilder::CreateReceiveCommandResponse(TEXT("SERVER_WORKER_FORWARD_SPAWN_REQUEST_COMMAND"), RequestId),
-				Op.span_id, 1);
+				EventTracer->GetSpanForRequestId(CommandResponseOp.request_id).GetConstId(), 1);
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:187]: Do we ever use anything other than `1` for the linear trace Id? We could add a default

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/ComponentUpdateEventTracingTest.cpp [hunk 19]
```diff
@@ -7,7 +7,7 @@ AComponentUpdateEventTracingTest::AComponentUpdateEventTracingTest()
 	Author = "Matthew Sandford";
 	Description = TEXT("Test checking the component update trace events have appropriate causes");
 
-	FilterEventNames = { ComponentUpdateEventName, ReceiveOpEventName, MergeComponentUpdateEventName };
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/ComponentUpdateEventTracingTest.cpp:10]: How come the removal of `MergeComponentUpdateEventName`?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/ProcessCrossServerRPCEventTracingTest.cpp [hunk 20]
```diff
@@ -0,0 +1,51 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "ProcessCrossServerRPCEventTracingTest.h"
+
+AProcessCrossServerRPCEventTracingTest::AProcessCrossServerRPCEventTracingTest()
+{
+	Author = "Danny Birch";
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/ProcessCrossServerRPCEventTracingTest.cpp:7]: "Big Man Danny B"*

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/ProcessCrossServerRPCEventTracingTest.cpp (hunk 11 above)
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/ProcessCrossServerRPCEventTracingTest.cpp:37]: Que?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/ProcessCrossServerRPCEventTracingTest.cpp [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+		// 		else // EventName == ReceiveRPCEventName
+		// 		{
+		// 			if (!CheckEventTraceCause(SpanIdString, { ReceiveOpEventName }, true))
+		// 			{
+		// 				EventsFailed++;
+		// 			}
+		// 		}
+	}
+
+	bool bSuccess = EventsTested > 0 && EventsFailed == 0;
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/ProcessCrossServerRPCEventTracingTest.cpp:46]: How come not a specific number for `EventsTested`?

**Comment by joshuahuburn** [comment]
Lovin it, couple comments and questions

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/ProcessCrossServerRPCEventTracingTest.cpp (hunk 21 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/ProcessCrossServerRPCEventTracingTest.cpp:46]: It's a timed scenario. If the time changes the number of events changes. Would be brittle if we had a set number.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/ComponentUpdateEventTracingTest.cpp (hunk 19 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/ComponentUpdateEventTracingTest.cpp:10]: MergeComponentUpdateEvent was an old event that does not exist any more.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/EventTracingTest.cpp (hunk 10 above)
Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/EventTracingTest.cpp:177]: Ah, I have adjusted the map to run with 2_1 zoning. Think I should add another map? It's nice to confirm that single client tracing does also function correctly with zoning but maybe it's a test for another place. Maybe a zoning map with the same tests in + the cross-server stuff?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp (hunk 15 above)
Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp:32]: Good question, it's not even 64-bit because INT_MAX is often 32-bit and how random is FRand (internally rand()) for uniqueness? What if it was seeded poorly?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 18 above)
Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:187]: 1 is the count of spans in the span* pointer, I will add your suggestion of commenting the args though so it's easier to spot

File: SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp (hunk 15 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp:32]: Mmmm. Seems like a weakness in our system for a random crash

Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCSender.cpp:32]: Yeah I wonder if a better approach might be N-bits of worker ID and a counter?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/UserProcessCrossServerPropertyEventTracingTest.cpp (hunk 12 above)
Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/EventTracingTests/UserProcessCrossServerPropertyEventTracingTest.cpp:23]: hmm this is the receive > user so we want to test UserReceiveCrossServerPropertyEventName causes below?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp [hunk 22]
```diff
@@ -518,9 +518,9 @@ void ActorSystem::ComponentUpdated(const Worker_EntityId EntityId, const Worker_
 
 	if (EventTracer != nullptr)
 	{
-		TArray<FSpatialGDKSpanId> CauseSpanIds = EventTracer->GetSpansForComponent(EntityComponentId(EntityId, ComponentId));
+		TArray<FSpatialGDKSpanId> CauseSpanIds = EventTracer->GetAndConsumeSpansForComponent(EntityComponentId(EntityId, ComponentId));
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:521]: Nit: PopSpansForComponent?

Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:521]: I guess it does not remove the spans immediately... ehh whatever you like

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp [hunk 23]
```diff
@@ -256,16 +269,36 @@ void SpatialEventTracer::UpdateComponent(const Worker_ComponentUpdateOp& Op, con
...
 	{
 		return {};
 	}
+	EntityComponentsConsumed.Push(Id); // Consume on frame boundary instead, as these can have multiple uses.
 	return *StoredSpanIds;
 }
 
+FSpatialGDKSpanId SpatialEventTracer::GetAndConsumeSpanForRequestId(Worker_RequestId RequestId)
+{
+	const FSpatialGDKSpanId* SpanId = RequestSpanIds.Find(RequestId);
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp:292]: Just interested to know if this ever actually starts to increase in size. I would hope we consume each request id we receive but might be good to check?

**Review by MatthewSandfordImprobable — APPROVED** [review]

**Review by joshuahuburn — APPROVED** [review]
Good from me <3

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp (hunk 23 above)
Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp:292]: Will check!

Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp:292]: Was all good - didn't seem to queue up when tracing cross-server!
